internal/controller/replication.storage/volumegroupreplication_controller.go

```diff
@@ -33,21 +57,536 @@ type VolumeGroupReplicationReconciler struct {
 	Scheme *runtime.Scheme
 }
 
-//+kubebuilder:rbac:groups=replication.storage.openshift.io,resources=volumegroupreplications,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=replication.storage.openshift.io,resources=volumegroupreplications,verbs=get;list;watch;update;patch
 //+kubebuilder:rbac:groups=replication.storage.openshift.io,resources=volumegroupreplications/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=replication.storage.openshift.io,resources=volumegroupreplications/finalizers,verbs=update
+//+kubebuilder:rbac:groups=replication.storage.openshift.io,resources=volumegroupreplicationclasses,verbs=get;list;watch
+//+kubebuilder:rbac:groups=replication.storage.openshift.io,resources=volumegroupreplicationcontents,verbs=get;list;watch;create;update;delete
+//+kubebuilder:rbac:groups=replication.storage.openshift.io,resources=volumereplications,verbs=get;list;watch;create;update;delete
+//+kubebuilder:rbac:groups=replication.storage.openshift.io,resources=volumereplications/status,verbs=get;list
+//+kubebuilder:rbac:groups=core,resources=persistentvolumeclaims;persistentvolumes,verbs=get;list;watch;update
+//+kubebuilder:rbac:groups=core,resources=persistentvolumeclaims/finalizers,verbs=update
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
 func (r *VolumeGroupReplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx, "Request.Name", req.Name, "Request.Namespace", req.Namespace)
```

Comment:
This function is now so large, it is difficult to see what it does. Can you document the steps that are taken, and split some of those tasks out in helper functions?

Reply:
Almost all the functions are split into helper functions, it seems big becuase there are few validations and CRs that are created in the Reconcile loop.

Here's the steps that are performed:
- Fetch and validate the VGRClass CR
- Fetch the matching PVCs based on the selector provided in the VGR CR, and check if they are already bounded to a CSI volume and the driver matches the driver provided in the VGRClass CR.
- Annotate to the PVCs with owner and add the VGR finalizer to them.
- Add the label selector to the VGR annotation, such that the PVC triggering a reconcile can fetch the VGR to reconcile
- Create the VGRContent with the PVs list fetched above, add VGR name/namespace as the annotation to it
- Wait for the volumes to be grouped, and the VGRContent to be updated with the group handle
- Then, create the VR CR and add VGR name/namespace as the annotation to it
- Update the VGR status with the VR status.

In case of deletion:
- Remove the owner annotations and finalizers from the PVC
- Check if VR exists, then delete
- Check if VGRContent exists, then delete
- Remove VGR finalizer <- This won't happen until the dependent VR and VRContent is deleted. Validated using owner annotations set in both the dependent CRs.

That's all that is being done.

Reply:
I have a doc as well, that have these steps documented: https://docs.google.com/document/d/1fOQPSpxhnQcW0xb0LUaUXHJv707VYqd7QXEcWFN5m5w/edit?usp=sharing

Reply:
It would be great to have the list of steps included in the description of the function. When someone looks at this function in 3 months time, nobody will know that it is explained in this PR or some google doc.

Reply:
Added the steps as a comment above the the `Reconcile` function definition.